[2기-P 이용훈] JPA 게시판 구현

YHLEE9753 opened the pull request:
SpringDataJpa를 설정한다.
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
JPA 게시판 과제 구현입니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
### **SpringDataJPA 를 설정한다.**

- [x] SpringDataJpa를 설정한다.
- datasource : h2 

- [x] **엔티티를 구성한다**

- 회원(User)
    - id (PK) (auto increment)
    - name
    - age
    - hobby
    - **created_at**
    - **created_by**
- 게시글(Post)
    - id (PK) (auto increment)
    - title
    - content
    - **created_at**
    - **created_by**
- 회원과 게시글에 대한 연관관계를 설정한다.
    - 회원과 게시글은 1:N 관계이다.
- 게시글 Repository를 구현한다. (PostRepository)

- [x] **API를 구현한다.**

- 게시글 조회
    - 페이징 조회 (GET "/posts")
    - 단건 조회 (GET "/posts/{id}")
- 게시글 작성 (POST "/posts")
- 게시글 수정 (PATCH "/posts/{id}")

- [x] **REST-DOCS를 이용해서 문서화한다.**

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

엔티티 Dto 변환 과정에서 맵핑 관계를 그대로 살리는 방식으로 코드를 작성하니 리소스 소모가 너무 큰거 같았습니다. 코드에서 질문드리겠습니다

Audit 설정에서 createdBy 를 시도할려고 했는데 실패했습니다. 좀더 소스를 찾아서 공부하도록 하겠습니다.

~~controller 에서 save update 에서 404 에러가 발생하였습니다. 해결이 안되서 일단 push 부터 진행하였습니다~~

너무 늦게 올려 죄송합니다.. ㅠㅠ 더욱더 열심히하겠습니다.

YHLEE9753 committed:
init: gitignore 설정

YHLEE9753 committed:
init: 프로젝트 셋팅
- application.yaml h2 db 연결
- gradle 설정

YHLEE9753 committed:
feat : JpaAudit 설정

YHLEE9753 committed:
feat : Post, User 엔티티 생성

YHLEE9753 committed:
feat : Post, User repository 생성

YHLEE9753 committed:
feat : Post, User Dto 생성

YHLEE9753 committed:
feat : Post, User Service 생성

YHLEE9753 committed:
feat : Post, User Service test 코드 작성

YHLEE9753 committed:
feat : Post, User converter 생성

YHLEE9753 committed:
feat : Rest Controller 생성

YHLEE9753 committed:
feat : Controller 테스트 코드 작성
- rest docs 적용

YHLEE9753 commented on src/main/java/com/kdt/board/global/config/PostAuditorAware.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public Optional<String> getCurrentAuditor() {
+        // 구현 필요
+        return Optional.empty();
```
createdby 를 audit 으로 설정할려고 했는데 좋은 소스를 찾지 못해 적용시키지 못했습니다. 추후 추가해보겠습니다

YHLEE9753 commented on src/main/java/com/kdt/board/domain/dto/PostDto.java:
```diff
@@ -0,0 +1,33 @@
+            Long id,
+            String title,
+            String content,
+            UserDto.Response user,
+            LocalDateTime createdAt,
```
이번 실습에서는 프로젝트상 의미있는 매핑 보다는 다양한 매핑을 시도해 볼려고 했습니다. 실제 다양한 변환 코드들을 보니 매핑 객체를 DTO 로 가지고 있는 경우가 많아 직접 시도해 보았습니다.

YHLEE9753 commented on src/main/java/com/kdt/board/domain/converter/Converter.java:
```diff
@@ -0,0 +1,59 @@
+        return Post.builder()
+                .title(postDto.title())
+                .content(postDto.content())
+                .user(convertService.entityFindById(postDto.userId()))
```
내부에 다른 객체의 Dto 를 가지게 코드를 짜다 보니 dto 에서 entity 변환시 직접 객체를 아이디로 찾아 집어 넣게 되었습니다. 코드 작성을 해보니 리소스 소모가 너무 큰거 같습니다. 실제 실무에서도 이런 방식으로 코드가 작성되는지 궁금합니다.

YHLEE9753 committed:
feat : 404 에러 해결 - url 변경
- id 값 확인용 dto 추가
- createdBy 어노테이션 수정 - nullable -> true
- test url 경로 추가(/post)

ordilov commented on src/test/java/com/kdt/board/domain/service/PostServiceTest.java:
```diff
@@ -0,0 +1,108 @@
+    Long savedUserId;
+    Long savedPostId;
+
+    @AfterEach
```
`@Transactional` 을 안 걸고 매번 초기화를 해주는 이유가 따로 있을까요?

ordilov commented on src/test/java/com/kdt/board/domain/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,220 @@
+        savedPostId = postService.save(postDto);
+
+        // then
+        assertThat(postRepository.count()).isEqualTo(1L);
```
테스트를 준비하기 위한 `@BeforeEach`에서 검증을 해주신 이유가 있을까요?

YHLEE9753 replied:
save test 코드를 재사용하다보니 불필요한 코드가 존재하네요 제거하도록 하겠습니다!

ordilov commented on src/test/java/com/kdt/board/domain/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,220 @@
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(saveDto))
+                )
+                .andExpect(status().isOk())
```
위에서 response 필드들을 작성해주신 것처럼 요청할 때 saveDto를 위해서 어떤 필드들이 필요한지도 문서에 작성되었으면 좋았을 것 같습니다!

YHLEE9753 replied:
작성하도록 하겠습니다!~

ordilov commented on src/main/java/com/kdt/board/domain/converter/ConvertService.java:
```diff
@@ -0,0 +1,22 @@
+    private final UserRepository userRepository;
+    private final PostRepository postRepository;
+
+    @Transactional
```
조회하는 메소드에 `@Transactional` 어노테이션을 사용해주신 이유가 있을까요?

YHLEE9753 replied:
db 접근 메서드이기때문에 사용하였습니다!

ordilov commented on src/main/java/com/kdt/board/domain/controller/PostController.java:
```diff
@@ -0,0 +1,61 @@
+        return ApiResponse.ok(updatedDto);
+    }
+
+    @DeleteMapping("/delete/{id}")
```
delete 메소드를 사용하는 것으로 해당하는 id의 post를 지운다는 걸 나타낼 수 있을 것 같아요!
path 상에 delete를 다시 명시해주신 이유가 있을까요?

YHLEE9753 replied:
rest API 디자인 가이드를 보니 적지 말라고 적혀있네요!! 인후님 덕분에 좀더 깊이 있게 공부할 수 있었습니다!

ordilov commented on src/main/java/com/kdt/board/domain/controller/PostController.java:
```diff
@@ -0,0 +1,61 @@
+        return ApiResponse.ok(posts);
+    }
+
+    @PostMapping("/save")
```
업데이트를 할 때 put 메소드가 있는 걸 봤을 때 post 메소드는 새로 저장할 때 쓰는 걸로 구분할 수 있을 것 같아요.
그럼 /save 없이도 새로 생성한다는 걸 나타낼 수 있을 것 같은데 따로 명시해주신 이유가 있을까요?

YHLEE9753 replied:
이 부분도 rest API 디자인 가이드에 따라 동일하게 제거하였습니다!

ordilov reviewed (approved):
용훈님~ 과제하시느라 고생 많으셨습니다!! 제가 잘못 요청드린 리뷰에도 꼼꼼하게 봐주시고 리뷰 달아주셔서 너무 감사했습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/controller/PostController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @GetMapping("/{id}")
+    public ApiResponse<PostDto.Response> getOne(
```
`getOne` 메소드 보다는 좀 더 구체적인 이름으로 변경하는 것은 어떤가요?

YHLEE9753 replied:
findById 를 통해 id 값을 통해 Dto 를 가져온다고 하는 것이 좀 더 구체적일것 같습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/controller/PostController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @PutMapping("/update")
+    public ApiResponse<PostDto.Response> update(
+            @RequestBody PostDto.Update dto
```
`@Valid` 와 같이 값을 검증할 수 있는 로직이 있으면 좋을 것 같습니다.

YHLEE9753 replied:
validation 공부 후 추가하였습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/controller/api/ApiResponse.java:
```diff
@@ -0,0 +1,29 @@
+
+public class ApiResponse<T>{
+    @JsonFormat(shape = JsonFormat.Shape.NUMBER)
+    private int statusCode;
```
`StatusCode` 를 Json 필드로 제공하는 이유를 알 수 있을까요 ??

YHLEE9753 replied:
필드값중 1개만 @JsonFormat 을 설정하면 @JsonFormat 을 적용시키지 않은 다른 필드에 값이 추가되지 않았습니다.(Body 에 값이 나타나지 않는 문제를 발견) 이런 문제를 해결하기 위해 모든 필드에 @JsonFormat 을 설정하였습니다.

hoon-space commented on src/main/java/com/kdt/board/domain/controller/api/ApiResponse.java:
```diff
@@ -0,0 +1,29 @@
+    private int statusCode;
+    @JsonFormat(shape = JsonFormat.Shape.OBJECT)
+    private T data;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
+    private LocalDateTime serverDatetime;
```
`serverDatetime` 이 필요한가요 ??

YHLEE9753 replied:
필요하지 않을 거 같습니다. @JsonFormat 으로 패턴과 timezone 설정이 가능하다는것을 실습하기 위해 따로 제거하지 않았습니다.

hoon-space commented on src/main/java/com/kdt/board/domain/controller/api/ApiResponse.java:
```diff
@@ -0,0 +1,29 @@
+        return new ApiResponse<>(200, data);
+    }
+
+    public static <T> ApiResponse<T> fail(int statusCode, T data){
+        return new ApiResponse<>(statusCode, data);
+    }
```
`data` 필드에 에러 메시지가 담기는 건가요 ?

YHLEE9753 replied:
넵 !! `@RestControllerAdvice` 를 사용하는 `GlobalExceptionHandler` 클래스에서 fail 함수를 사용하여 에러 메시지가 담기도록 하였습니다!!

hoon-space commented on src/main/java/com/kdt/board/domain/converter/ConvertService.java:
```diff
@@ -0,0 +1,22 @@
+@Service
+@RequiredArgsConstructor
+public class ConvertService {
+    private final UserRepository userRepository;
+    private final PostRepository postRepository;
```
각 도메인 마다 convert 로직이 분리 되는 것이 좋을 것 같은데, 용훈님 생각은 어떤가요 ??

YHLEE9753 replied:
user 과 post 가 서로 동시에 convert 하고 있기 때문에 `ConvertService `를 하나로 사용하였습니다. 하지만 클래스 책임 및 확장성을 고려할때는 분리하는것이 좋을 거 같습니다.

hoon-space commented on src/main/java/com/kdt/board/domain/converter/ConvertService.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
```
convert 로직이 서비스인가요 ??

YHLEE9753 replied:
`@Component `가 더 적절한 거 같습니다 변경하였습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/converter/Converter.java:
```diff
@@ -0,0 +1,59 @@
+public class Converter {
+    private final ConvertService convertService;
+
+    // dto -> entity
```
주석이 필요할까요 ??

YHLEE9753 replied:
제거 하였습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/dto/PostDto.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.time.LocalDateTime;
+
+public class PostDto {
```
레이어 별로 Dto 가 분리되면 더 좋지 않을까요 ??

hoon-space commented on src/main/java/com/kdt/board/domain/model/Post.java:
```diff
@@ -0,0 +1,52 @@
+        this.content = content;
+    }
+
+    // 연관 관계편의 메서드
```
비즈니스적인 로직이 아니라면, 해당 주석은 필요 없을 것 같습니다.

YHLEE9753 replied:
제거 하였습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/model/User.java:
```diff
@@ -0,0 +1,46 @@
+        this.hobby = hobby;
+    }
+
+    public void update(String name, int age, String hobby) {
```
무엇을 업데이트하는지 좀 더 명확하게 네이밍하면 좋을 것 같아요 ~

YHLEE9753 replied:
`updateUser`  로 자신을 update 한다고 표현하였습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/service/PostService.java:
```diff
@@ -0,0 +1,57 @@
+        return entity.getId();
+    }
+
+    @Transactional
+    public PostDto.Response findById(Long id) {
```
`@Transactional` 애노테이션의 readOnly 옵션을 사용해봐도 좋을 것 같아요

YHLEE9753 replied:
수정하였습니다!!

hoon-space commented on src/main/java/com/kdt/board/domain/service/PostService.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Transactional
+    public long count() {
+        return postRepository.count();
```
전체를 count 하는 로직은 성능상 최대한 자제하는 것이 좋습니다.

YHLEE9753 replied:
test 상에서만 사용하고 실제 서비스에서는 사용하는것을 자제하도록 하겠습니다! 가능하다면 exists 를 사용하겠습니다.

hoon-space commented on src/main/java/com/kdt/board/domain/service/UserService.java:
```diff
@@ -0,0 +1,59 @@
+    @Transactional
+    public UserDto.Response findById(Long id) {
+        return userRepository.findById(id)
+                .map(converter::convertUserDto)
+                .orElseThrow(() -> new LoadEntityException("Entity 를 불러오는 중 예외 발생"));
+    }
```
해당 예외가 발생했을 때, 어떤 statusCode 를 보내줬나요 ?

YHLEE9753 replied:
`@RestControllerAdvice` 에서 ` HttpStatus.NOT_FOUND `를 보내주었습니다!

hoon-space reviewed (commented):
과제 하시느라 고생 많으셨습니다.

수정이 필요한 부분 코멘트드렸고, 확인해보세요 ~

YHLEE9753 committed:
feat : url 설정(rest api 디자인 적용)

YHLEE9753 commented on src/main/java/com/kdt/board/domain/dto/PostDto.java:
```diff
@@ -0,0 +1,39 @@
+    ) {
+    }
+
+    public record checkingId(
```
소문자로 적었다니.... 대문자로 바꾸겠습니다ㅠㅠ

YHLEE9753 committed:
feat : record class 소문자 네이밍 대문자로 변경

YHLEE9753 commented on src/main/java/com/kdt/board/domain/dto/PostDto.java:
```diff
@@ -0,0 +1,39 @@
+    ) {
+    }
+
+    public record Update(
```
네이밍에 Request Response 까지 적어주는것이 훨씬 더 세부적인거 같습니다!

YHLEE9753 committed:
feat : record class 네이밍 변경(response request 추가)

YHLEE9753 committed:
feat : RequestBody 검증 로직 구현
- @Valid 설정
- RestControllerAdvice 설정

YHLEE9753 committed:
feat : 컨벤션 수정, 클래스 책임분할
- 네이밍 컨벤션 수정
- 클래스 분리
- handleEntityNotFoundException 예외처리 추가
- 필요없는 주석 제거
- UserConvertService @Service -> @Component 수정

jungyoonoh commented on src/main/java/com/kdt/board/global/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,10 @@
+
+@Configuration
+@EnableJpaAuditing
+public class JpaAuditingConfig {
```
이 configuration을 별도의 클래스로 빼서 작업하신 이유가 있을까요?

jungyoonoh commented on src/test/java/com/kdt/board/BoardApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+class BoardApplicationTests {
+
+	@Test
+	void contextLoads() {
```
이 함수는 어떤 목적으로 작성된 것인가요?

jungyoonoh reviewed (approved):
용훈님 수고하셨습니다!
리뷰가 늦어 죄송합니다 ㅠㅠ 
멘토님과 인후님이 작성 잘 해주셔서 그 부분들만 보시면 좋을 것 같습니다!

hoon-space commented on src/main/java/com/kdt/board/domain/converter/UserConvertService.java:
```diff
@@ -0,0 +1,21 @@
+
+@Component
+@RequiredArgsConstructor
+public class UserConvertService {
```
클래스 이름은 Service 인데 애노테이션은 `@Component` 네요

hoon-space commented on src/main/java/com/kdt/board/domain/converter/UserConvertService.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Transactional
+    public User entityFindById(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new NotFoundException("Entity 를 불러오는 중 예외 발생"));
```
`Entity 를 불러오는 중 예외 발생` 이라는 메시지 보다는 Entity 를 찾을 수 없다는 의미를 담는게 좀 더 명확해보입니다.

hoon-space commented on src/main/java/com/kdt/board/domain/converter/UserConvertService.java:
```diff
@@ -0,0 +1,21 @@
+public class UserConvertService {
+    private final UserRepository userRepository;
+
+    @Transactional
```
읽기 전용 트랜잭션인 것 같습니다. 

변경감지와 같은 기능을 할 필요가 없기 때문에 readOnly 로 변경하는 것이 좋아보여요

hoon-space commented on src/main/java/com/kdt/board/domain/dto/PostDto.java:
```diff
@@ -0,0 +1,44 @@
+import javax.validation.constraints.NotNull;
+import java.time.LocalDateTime;
+
+public class PostDto {
```
저는 현업에서 자바 17을 쓰지 않아서 저도 잘 모르는 건데 

자바에서 클래스안에 클래스를 만들 때, static 을 붙이지 않으면 메모리 낭비 + GC 가 제대로 동작하지 않습니다.
여기 코드를 보면 Record 로 여러 클래스를 생성했는데, GC 에 문제가 없는지 확인해봐야 될 것 같아요.

그리고 굳이 중첩해서 만들 필요가 있는지도 고민해보세요 ~ 
생명 주기를 같이 하는 클래스가 아니면 내부에 정의할 필요는 없다고 생각합니다. 

참고 
https://tecoble.techcourse.co.kr/post/2020-11-05-nested-class/
https://johngrib.github.io/wiki/java-inner-class-may-be-static/

hoon-space commented on src/main/java/com/kdt/board/domain/model/Post.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void changeUser(User user) {
+        if (Objects.nonNull(this.user)) {
+            this.user.getPosts().remove(this);
```
Object.nonNull( ).XXX( ) 이런식으로 사용할 의도가 아니라면 
if (x != null) { } 이 좀 더 심플해보입니다.

단순 null 체크이니 Object 를 불러와서 객체를 wrapping 할 필요는 없어보입니다.

hoon-space commented on src/test/java/com/kdt/board/domain/service/PostServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+
+@SpringBootTest
+class PostServiceTest {
+    @Autowired
+    private PostService postService;
```
테스트 코드이니 굳이 private 을 붙이지 않아도 됩니다.

hoon-space commented on src/test/java/com/kdt/board/domain/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,215 @@
+                .andExpect(status().isOk())
+                .andDo(print())
+                .andDo(document("post-findAll",
+                                responseFields(
+                                        fieldWithPath("statusCode").type(NUMBER).description("상태코드"),
+                                        fieldWithPath("serverDatetime").type(STRING).description("응답시간"),
+                                        fieldWithPath("data").type(OBJECT).description("데이터"),
+                                        fieldWithPath("data.content").type(ARRAY).description("데이터목록"),
+                                        fieldWithPath("data.content[].id").type(NUMBER).description("게시판아이디"),
+                                        fieldWithPath("data.content[].title").type(STRING).description("게시판제목"),
+                                        fieldWithPath("data.content[].content").type(STRING).description("게시판내용"),
+                                        fieldWithPath("data.content[].createdAt").type(STRING).description("content createdAt"),
```
파라미터가 너무 많다면, private 메서드로 분리해보는 것도 좋습니다.

hoon-space reviewed (commented):
수정한 부분에서 아쉬운 점 코멘트 남겼습니다.

추가로 아래 두 가지 부분도 체크해주세요 ~

1) 테스트 코드
통합테스트로만 작성한 것 같습니다. 
목킹 라이브러리를 사용해서 계층별 격리를 시킨 유닛테스트를 작성해보도록 노력해보는 것도 중요합니다.


2) .gitIgnore 
불필요한 파일이 같이 올라가는 것 같습니다. 아래 첨부한 이미지에서 정말 필요한 파일인지 고민해보시고, 필요 앖다면 제거해보세요 ~

![image](https://user-images.githubusercontent.com/101714618/170505147-865cc02a-6300-4e07-b879-858c1d794920.png)

mentor-hoon commented:
용훈님 기말고사 끝나고 다시 PR 올려주세요 !!!
일단 해당 PR 은 머지하겠습니다. !!
